src/engine.rs

```diff
@@ -381,8 +381,27 @@ impl Reedline {
                     reedline_events.push(ReedlineEvent::Resize(x, y));
                 }
 
+                let mut last_edit_commands = None;
                 for event in crossterm_events.drain(..) {
-                    reedline_events.push(self.edit_mode.parse_event(event))
+                    match (&mut last_edit_commands, self.edit_mode.parse_event(event)) {
+                        (None, ReedlineEvent::Edit(ec)) => {
+                            last_edit_commands = Some(ec);
+                        }
+                        (None, other_event) => {
+                            reedline_events.push(other_event);
+                        }
+                        (Some(ref mut last_ecs), ReedlineEvent::Edit(ec)) => {
+                            last_ecs.extend(ec);
+                        }
+                        (ref mut a @ Some(_), other_event) => {
+                            reedline_events.push(ReedlineEvent::Edit(a.take().unwrap()));
+
+                            reedline_events.push(other_event);
+                        }
+                    }
+                }
+                if let Some(ec) = last_edit_commands {
+                    reedline_events.push(ReedlineEvent::Edit(ec));
```

Comment:
This is certainly not the most readable piece of code. `itertools::Itertools::coalesce` would be probably a nicer replacement (Writing it out myself might be sign of premature optimization)